Bricks/sync/scope_owned.h

```diff
@@ -0,0 +1,388 @@
+/*******************************************************************************
+The MIT License (MIT)
+
+Copyright (c) 2016 Dmitry "Dima" Korolev, <dmitry.korolev@gmail.com>.
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+// The synchronization primitive to leverage RAII and enable creating scope-based long-lived objects.
+//
+// Philosophy:
+// * The scope in which the object has been created is the master owner of this object; others are followers.
+// * Upon leaving the master scope, the synchronization primitive waits until all its followers terminate.
+// * Followers are under no obligation to terminate sooner or later, but they get signaled to wrap up.
+//
+// Implementation:
+//
+// * ScopeOwnedByMe<T> master(T_construction_parameter1, ...);
+//   Creates `master`, which will live only within its creation scope.
+//
+// * ScopeOwnedBySomeoneElse<T> follower(master, [&signal]() { signal.SignalTermination(); });
+//   Initializes `follower`, likely `std::move()` it into a different thread.
+//   The scope of the `follower` object will be strictly inside the scope of `master`.
```

Comment:
I think you should say "lifetime" here instead of "scope". Scope refers to a syntactic region.